### PR TITLE
Create datasource API

### PR DESCRIPTION
**Pull request by vamsi-amazon**
### Description
This PR covers Create datasource REST API.
 

#### Create datasource [REST API]

Screen Shot 2023-03-06 at 6 09 58 PM

#### Get datasource along with storage engine and other artifacts [Internal use case only]

Screen Shot 2023-03-06 at 6 13 39 PM

### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by ps48 — DISMISSED** [review]

**Comment by dai-chen** [comment]
Any special reason this has to merge to 2.x first?

**Review by dai-chen — COMMENTED** [review]
I have some high level questions because this PR seems have more influence than it looks:

1. Is `.ql-data-sources` switch SQL plugin to stateful? If the index is red for some reason, want to confirm if SQL plugin still work as before? At lease for query on default OpenSearch data source.
2. Data source related code and dependencies keep growing, should we move all of them to a new `datasource` module. Because as I understand, `core` only depends on resolved `DataSource` or `Table` actually.

File: legacy/src/main/java/org/opensearch/sql/legacy/metrics/MetricName.java [hunk 1]
```diff
@@ -27,7 +27,10 @@ public enum MetricName {
     PPL_REQ_TOTAL("ppl_request_total"),
     PPL_REQ_COUNT_TOTAL("ppl_request_count"),
     PPL_FAILED_REQ_COUNT_SYS("ppl_failed_request_count_syserr"),
-    PPL_FAILED_REQ_COUNT_CUS("ppl_failed_request_count_cuserr");
+    PPL_FAILED_REQ_COUNT_CUS("ppl_failed_request_count_cuserr"),
+    DATASOURCE_REQ_COUNT("datasource_request_count"),
+    DATASOURCE_FAILED_REQ_COUNT_SYS("datasource_failed_request_count_syserr"),
+    DATASOURCE_FAILED_REQ_COUNT_CUS("datasource_failed_request_count_cuserr");
```
[the remark this answers is not in the document]
Comment (vamsi-amazon) [reply legacy/src/main/java/org/opensearch/sql/legacy/metrics/MetricName.java:33]: > Any special reason this has to merge to 2.x first?

I need this for pen testing on 2.7 version.  I need to resolve conflicts with main. I will raise backport PR once that is done.

[the remark this answers is not in the document]
Comment (vamsi-amazon) [reply legacy/src/main/java/org/opensearch/sql/legacy/metrics/MetricName.java:33]: Will take a look into this.

File: plugin/src/main/java/org/opensearch/sql/plugin/model/CreateDataSourceActionRequest.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    super(in);
+  }
+
+  public CreateDataSourceActionRequest(DataSourceMetadata dataSourceMetadata) {
+    this.dataSourceMetadata = dataSourceMetadata;
+  }
+
+  @Override
+  public ActionRequestValidationException validate() {
+    return null;
```
[the remark this answers is not in the document]
Comment (vamsi-amazon) [reply plugin/src/main/java/org/opensearch/sql/plugin/model/CreateDataSourceActionRequest.java:37]: Auto generated. Null means there are no errors. I have added small validation here in upcoming PR, I will try to add it in the next revision itself.

**Comment by vamsi-amazon** [comment]
> I have some high level questions because this PR seems have more influence than it looks:
> 
> 1. There is no issue related to this main feature. I think we should create something and put it to our 2.7 or roadmap

I will add all the required issues.

> 2. The new `.ql-data-sources` index is shifting stateless SQL plugin to stateful? If the index is red for some reason, want to confirm if SQL plugin still work as before? At lease for query on default OpenSearch data source.

It should work for queries on default Opensearch Connector even in case of red `.ql-datasources` index. I will go through the code again and list if there are any scenarios where it breaks.

> 3. Data source related code and dependencies keep growing, should we move all of them to a new `datasource` module. Because as I understand, `core` only depends on resolved `DataSource` or `Table` actually.
I was thinking on similar lines whether this code is really required in core or some other place. will discuss with you and try to move this.

**Pull request closed without merging**